feat: Supply ParsingError with additional information

mykola-mokhnach-parloa opened the pull request:
The added info should help the client code to compile better error messages about possible templating issues

mykola-mokhnach-parloa committed:
feat: Supply ParsingError with additional information

mykola-mokhnach-parloa committed:
feat: Supply ParsingError with additional information

bakert commented:
Thanks for the PR!

I gave it a little run out with:

```
import pystache; print(pystache.render('Hi\n\n\n\n\n {{/person}}!', {'person': 'Mom'}))
```

I do agree that:

`pystache.parser.ParsingError: Section end tag mismatch at the position 19. Found {{/person}}, expected {{#person}}`

is a nicer error message than:

`pystache.parser.ParsingError: Section end tag mismatch: person != None` seems broadly as good to me as 

(Grammatical nit: it should be "at position 19" not "at the position 19", or maybe "at character 19".)

I am a little wary of the extra complication here, especially the rather magical looking stuff like this:

```
            exc_state = (
                states[0][0],
                states[0][1],
                states[0][2] or tag_key,
            )
```

I wonder if we want something more like this, elsewhere in parser.py, where these things are given nice names:

```
(
                    tag_type,
                    section_start_index,
                    section_key,
                    parsed_template,
                ) = states.pop()
```

It's also true that the tests didn't run successfully but it looks like that's some kind of bitrot in the process rather than a problem with this PR. I'll fix that up and re-run.

mykola-mokhnach-parloa committed:
Address comments

mykola-mokhnach-parloa committed:
Merge branch 'parse_err' of github.com:mykola-mokhnach-parloa/pystache into parse_err

# Conflicts:
#	pystache/parser.py

mykola-mokhnach-parloa commented:
> Thanks for the PR!
> 
> I gave it a little run out with:
> 
> ```
> import pystache; print(pystache.render('Hi\n\n\n\n\n {{/person}}!', {'person': 'Mom'}))
> ```
> 
> I do agree that:
> 
> `pystache.parser.ParsingError: Section end tag mismatch at the position 19. Found {{/person}}, expected {{#person}}`
> 
> is a nicer error message than:
> 
> `pystache.parser.ParsingError: Section end tag mismatch: person != None`
> 
> (Grammatical nit: it should be "at position 19" not "at the position 19", or maybe "at character 19".)
> 
> I am a little wary of the extra complication here, especially the rather magical looking stuff like this:
> 
> ```
>             exc_state = (
>                 states[0][0],
>                 states[0][1],
>                 states[0][2] or tag_key,
>             )
> ```
> 
> I wonder if we want something more like this, elsewhere in parser.py, where these things are given nice names:
> 
> ```
>                (
>                     tag_type,
>                     section_start_index,
>                     section_key,
>                     parsed_template,
>                 ) = states.pop()
> ```
> 
> It's also true that the tests didn't run successfully but it looks like that's some kind of bitrot in the process rather than a problem with this PR. I'll fix that up and re-run.

I've addressed your comments @bakert 
Exception properties have been changed to a data object to improve readability

mykola-mokhnach-parloa committed:
upd tag_type

bakert commented:
Thanks for the PR!

bakert commented:
This is now in v0.6.8 which is on pypi and is now the latest release.
